### PR TITLE
refactor: Remove all API helpers from the API prototype

### PR DESCRIPTION
**Pull request by MarcosSpessatto**
## Proposed changes (including videos or screenshots)

This PR removes all the API helpers from the API prototype and converts it to regular functions instead.
Jira: ARCH-934

### CODE REVIEW

File: apps/meteor/app/livechat/imports/server/rest/departments.ts [hunk 1]
```diff
@@ -28,8 +30,15 @@ API.v1.addRoute(
...
-			const { sort } = this.parseJsonQuery();
+			const { offset, count } = await getPaginationItems(this.queryParams);
+			const { sort } = await parseJsonQuery(
+				this.request.route,
+				this.userId,
+				this.queryParams,
+				this.logger,
+				this.queryFields,
+				this.queryOperations,
+			);
```
Comment (murtaza98) [inline apps/meteor/app/livechat/imports/server/rest/departments.ts:41]: I didn't like that we'd to pass these many props to this method which is getting used a lot of the times across the codebase. Considering all these props are already part of the generic API class, my suggestion would be to move this function (i.e. `parseJsonQuery` ), inside this API class. That way, on the endpoints, we'd only require to call `const { sort } = await this.parsedJsonQuery()`. 

Let me know you thoughts on this Marcos :)

Comment (MarcosSpessatto) [reply apps/meteor/app/livechat/imports/server/rest/departments.ts:41]: It makes sense, but the big problem is that it is a huge function, not sure if it is worth moving all the code inside the class (which is also a very large file), but I can do it as a wrapper maybe. Keep calling the external function, but having that class method as you suggest, wdyt?

**Review by murtaza98 — APPROVED** [review]

File: apps/meteor/app/api/server/helpers/getPaginationItems.ts [hunk 2]
```diff
@@ -1,21 +1,25 @@
...
-import { API } from '../api';
 
-API.helperMethods.set('getPaginationItems', function _getPaginationItems(this: any) {
-	const hardUpperLimitTest = settings.get<number>('API_Upper_Count_Limit');
-	const defaultCountTest = settings.get<number>('API_Default_Count');
+export async function getPaginationItems(params: { offset?: number; count?: number }): Promise<{
+	readonly offset: number;
+	readonly count: number;
+}> {
+	const hardUpperLimitTest = ((await Settings.findOneById('API_Upper_Count_Limit'))?.value as number) || 0;
```
[the remark this answers is not in the document]
Comment (KevLehman) [reply apps/meteor/app/api/server/helpers/getPaginationItems.ts:12]: We can keep using the other settings for now, while they're made async

File: apps/meteor/app/api/server/helpers/getPaginationItems.ts [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
+	const defaultCountTest = ((await Settings.findOneById('API_Default_Count'))?.value as number) || 0;
 
 	const hardUpperLimit = hardUpperLimitTest && hardUpperLimitTest <= 0 ? 100 : settings.get<number>('API_Upper_Count_Limit');
```
Comment (KevLehman) [inline apps/meteor/app/api/server/helpers/getPaginationItems.ts:15]: Like these

File: apps/meteor/app/api/server/helpers/parseJsonQuery.ts [hunk 4]
```diff
@@ -18,134 +19,133 @@ const warnFields =
...
-			}
+	}
+
+	// Verify the user's selected fields only contains ones which their role allows
+	if (typeof fields === 'object') {
+		let nonSelectableFields = Object.keys(API.v1.defaultFieldsToExclude);
+		if (route.includes('/v1/users.')) {
+			const getFields = async (): Promise<string[]> =>
+				Object.keys(
+					(await hasPermission(userId, 'view-full-other-user-info'))
```
Comment (KevLehman) [inline apps/meteor/app/api/server/helpers/parseJsonQuery.ts:84]: hasPermissionAsync

File: apps/meteor/app/api/server/v1/channels.ts [hunk 5]
```diff
@@ -517,7 +520,7 @@ API.v1.addRoute(
 		async post() {
 			const findResult = await findChannelByIdOrName({ params: this.bodyParams });
 
-			const user = this.getUserFromParams();
+			const user = await getUserFromParams(this.bodyParams as Record<string, any>);
```
Comment (KevLehman) [inline apps/meteor/app/api/server/v1/channels.ts:523]: 🤔 maybe getUserFromParams should accept the same type this.bodyParams is

File: apps/meteor/app/api/server/helpers/getLoggedInUser.ts [hunk 6]
```diff
@@ -1,28 +1,16 @@
...
-
 export async function getLoggedInUser(token: string, userId: string): Promise<Pick<IUser, '_id' | 'username'> | null> {
 	if (!token || !userId) {
 		return null;
 	}
-	return UsersRaw.findOne(
+	return Users.findOne(
 		{
 			'_id': userId,
 			'services.resume.loginTokens.hashedToken': Accounts._hashLoginToken(token),
```
Comment (KevLehman) [inline apps/meteor/app/api/server/helpers/getLoggedInUser.ts:12]: I think we have a query for this on the model 🤔 

But, btw, you can also use the same _hashLoginToken from account-utils package

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: apps/meteor/app/api/server/helpers/getPaginationItems.ts [hunk 7]
```diff
@@ -24,12 +28,14 @@ API.helperMethods.set('getPaginationItems', function _getPaginationItems(this: a
 		count = hardUpperLimit;
 	}
 
-	if (count === 0 && !settings.get('API_Allow_Infinite_Count')) {
+	const infiniteCount = (await Settings.findOneById('API_Allow_Infinite_Count'))?.value;
```
Comment (sampaiodiego) [inline apps/meteor/app/api/server/helpers/getPaginationItems.ts:31]: any special reason for this to not be a `settings.get` as the settings above?

File: apps/meteor/app/api/server/v1/teams.ts [hunk 8]
```diff
@@ -183,7 +184,7 @@ API.v1.addRoute(
 				return API.v1.unauthorized('error-no-permission-team-channel');
 			}
 
-			const { rooms } = this.bodyParams;
+			const { rooms } = this.bodyParams as unknown as { rooms: string[] };
```
Comment (sampaiodiego) [inline apps/meteor/app/api/server/v1/teams.ts:187]: you can improve the `check` above so you don't need this type casting, something like:
```ts
			check(
				this.bodyParams,
				Match.OneOf(
					Match.ObjectIncluding({
						teamId: String,
						rooms: [String],
					}),
					Match.ObjectIncluding({
						teamName: String,
						rooms: [String],
					}),
				),
			);
```

File: apps/meteor/app/api/server/v1/users.ts [hunk 9]
```diff
@@ -578,8 +582,8 @@ API.v1.addRoute(
 	'users.createToken',
 	{ authRequired: true },
 	{
-		post() {
-			const user = this.getUserFromParams();
+		async post() {
+			const user = await getUserFromParams(this.bodyParams as unknown as Record<string, any>);
```
Comment (rodrigok) [inline apps/meteor/app/api/server/v1/users.ts:586]: Unnecessary typecast

File: apps/meteor/app/api/server/v1/users.ts [hunk 10]
```diff
@@ -1089,13 +1093,14 @@ API.v1.addRoute(
 					method: 'users.setStatus',
 				});
 			}
+			const params = this.bodyParams as unknown as Record<string, any>;
```
Comment (rodrigok) [inline apps/meteor/app/api/server/v1/users.ts:1096]: Unnecessary typecast

File: apps/meteor/app/api/server/v1/users.ts [hunk 11]
```diff
@@ -1153,8 +1158,9 @@ API.v1.addRoute(
 	'users.getStatus',
 	{ authRequired: true },
 	{
-		get() {
-			if (this.isUserFromParams()) {
+		async get() {
+			const params = this.queryParams as unknown as Record<string, any>;
```
Comment (rodrigok) [inline apps/meteor/app/api/server/v1/users.ts:1162]: Unnecessary typecast

File: apps/meteor/app/api/server/v1/teams.ts (hunk 8 above)
Comment (rodrigok) [inline apps/meteor/app/api/server/v1/teams.ts:187]: Typecast is always a bad idea

File: apps/meteor/app/api/server/helpers/isUserFromParams.ts [hunk 12]
```diff
@@ -1,12 +1,14 @@
...
+	loggedInUser?: IUser,
+): boolean {
+	return Boolean(
 		(!params.userId && !params.username && !params.user) ||
-		(params.userId && this.userId === params.userId) ||
-		(params.username && this.user.username === params.username) ||
-		(params.user && this.user.username === params.user)
+			(params.userId && loggedInUserId === params.userId) ||
+			(params.username && loggedInUser?.username === params.username) ||
+			(params.user && loggedInUser?.username === (params.user as unknown as string)),
```
Comment (rodrigok) [inline apps/meteor/app/api/server/helpers/isUserFromParams.ts:12]: Typecast is always a bad idea

**Review by rodrigok — CHANGES_REQUESTED** [review]

File: apps/meteor/app/api/server/helpers/getUserFromParams.ts [hunk 13]
```diff
@@ -75,34 +53,9 @@ export async function getUserListFromParams(params: {
 
 	if (params.userIds || params.userId) {
 		users = await Users.findByIds(userListParam, { fields: { username: 1 } });
-	} else {
-		users = await Users.findByUsernamesIgnoringCase(userListParam, { fields: { username: 1 } });
+		return users.toArray();
 	}
+	users = await UsersSync.findByUsernamesIgnoringCase(userListParam, { fields: { username: 1 } });
```
Comment (sampaiodiego) [inline apps/meteor/app/api/server/helpers/getUserFromParams.ts:58]:
```suggestion
	users = await Users.findByUsernamesIgnoringCase(userListParam, { projection: { username: 1 } });
```

File: apps/meteor/app/api/server/helpers/getUserInfo.ts [hunk 14]
```diff
@@ -33,10 +42,10 @@ API.helperMethods.set('getUserInfo', function _getUserInfo(me: IUser) {
 		settings: {
 			profile: {},
 			preferences: {
-				...getUserPreferences(me),
+				...(await getUserPreferences(me)),
 				...userPreferences,
 			},
 		},
-		avatarUrl: getURL(`/avatar/${me.username}`, { cdn: false, full: true }),
+		avatarUrl: await getURL(`/avatar/${me.username}`, { cdn: false, full: true }),
```
Comment (sampaiodiego) [inline apps/meteor/app/api/server/helpers/getUserInfo.ts:49]:
```suggestion
		avatarUrl: getURL(`/avatar/${me.username}`, { cdn: false, full: true }),
```

**Review by sampaiodiego — CHANGES_REQUESTED** [review]

File: apps/meteor/app/api/server/helpers/getPaginationItems.ts (hunk 7 above)
Comment (MarcosSpessatto) [reply apps/meteor/app/api/server/helpers/getPaginationItems.ts:31]: Not really, I thought we should remove all settings.get in favor of the model, but let me undo the change.

File: apps/meteor/app/api/server/helpers/getLoggedInUser.ts [hunk 15]
```diff
@@ -1,28 +1,14 @@
...
 
-API.helperMethods.set('getLoggedInUser', function _getLoggedInUser(this: any) {
-	if (this.request.headers['x-auth-token'] && this.request.headers['x-user-id']) {
-		return Users.findOne({
-			'_id': this.request.headers['x-user-id'],
-			'services.resume.loginTokens.hashedToken': Accounts._hashLoginToken(this.request.headers['x-auth-token']),
-		});
-	}
-});
+	return Users.findOneByIdAndLoginToken(userId as string, Accounts._hashLoginToken(token as string), { projection: { username: 1 } });
```
[the remark this answers is not in the document]
Comment (MarcosSpessatto) [reply apps/meteor/app/api/server/helpers/getLoggedInUser.ts:13]: yes, it should be better, let me change

File: apps/meteor/app/api/server/helpers/getPaginationItems.ts [hunk 16]
```diff
@@ -2,20 +2,22 @@
...
 	const hardUpperLimit = hardUpperLimitTest && hardUpperLimitTest <= 0 ? 100 : settings.get<number>('API_Upper_Count_Limit');
 	const defaultCount = defaultCountTest && defaultCountTest <= 0 ? 50 : settings.get<number>('API_Default_Count');
-	const offset = this.queryParams.offset ? parseInt(this.queryParams.offset) : 0;
+	const offset = params.offset ? parseInt(String(params.offset || 0)) : 0;
 	let count = defaultCount;
 
 	// Ensure count is an appropriate amount
-	if (typeof this.queryParams.count !== 'undefined') {
-		count = parseInt(this.queryParams.count);
+	if (params.count) {
```
Comment (KevLehman) [inline apps/meteor/app/api/server/helpers/getPaginationItems.ts:19]: If count === 0 this will put defaultCount and not trigger the check on L29

**Review by KevLehman — CHANGES_REQUESTED** [review]

**Review by sampaiodiego — APPROVED** [review]

**Review by rodrigok — APPROVED** [review]

**Pull request merged**